File: fcompare.f90
! Take 2 plotfiles as input and compare them zone by zone for differences.
!
! For the comparison to take place, the grids must be identical.
!

program fcompare

  use f2kcli
  use bl_space
  use bl_error_module
  use bl_constants_module
  use bl_IO_module
  use plotfile_module
  use sort_d_module

  implicit none

  type(plotfile) :: pf_a, pf_b
  character (len=256) :: plotfile_a, plotfile_b
  integer :: unit_a, unit_b

  real(kind=dp_t), pointer :: p_a(:,:,:,:), p_b(:,:,:,:)

  integer :: lo_a(MAX_SPACEDIM), hi_a(MAX_SPACEDIM)
  integer :: lo_b(MAX_SPACEDIM), hi_b(MAX_SPACEDIM)

  integer :: flo_a(MAX_SPACEDIM), fhi_a(MAX_SPACEDIM)
  integer :: flo_b(MAX_SPACEDIM), fhi_b(MAX_SPACEDIM)

  real(kind=dp_t) :: dx_a(MAX_SPACEDIM), dx_b(MAX_SPACEDIM)

  integer :: nboxes_a, nboxes_b

  integer :: n_a, n_b
  integer, allocatable :: ivar_b(:)

  real(kind=dp_t), allocatable :: aerror(:), rerror(:)

  logical, allocatable :: has_nan_a(:), has_nan_b(:)
  logical :: any_nans

  integer :: norm

  integer :: narg, farg
  character (len=256) :: fname

  integer :: i, j
  integer :: ii, jj, kk

  integer ir_a, ir_b

  integer :: itest

  real(kind=dp_t) :: global_error
  real(kind=dp_t) :: pa, pb, pd

  integer :: dm
  type(box) :: bx_a, bx_b


  !---------------------------------------------------------------------------
  ! process the command line arguments

  narg = command_argument_count()

  ! defaults
  norm = 0
  plotfile_a = ""
  plotfile_b = ""

  farg = 1
  do while (farg <= narg)
     call get_command_argument(farg, value = fname)
     
     select case (fname)

     case ('--infile1')
        farg = farg + 1
        call get_command_argument(farg, value = plotfile_a)

     case ('--infile2')
        farg = farg + 1
        call get_command_argument(farg, value = plotfile_b)

     case ('-n','--norm')
        farg = farg + 1
        call get_command_argument(farg, value = fname)
        read(fname, *) norm

     case default
        exit

     end select
     farg = farg + 1
  enddo

  if (len_trim(plotfile_a) == 0) then
     call get_command_argument(farg, value = plotfile_a)
     farg = farg + 1
  endif

  if (len_trim(plotfile_b) == 0) then
     call get_command_argument(farg, value = plotfile_b)
     farg = farg + 1
  endif

  if (len_trim(plotfile_a) == 0 .OR. len_trim(plotfile_b) == 0) then
     print *, " "
     print *, "Compare two plotfiles, zone by zone, to machine precision"
     print *, "and report the maximum absolute and relative errors for each"
     print *, "variable."
     print *, " "
     print *, "usage:"
     print *, "   fcompare --infile1 file1 --infile2 file2"
     print *, ' '
     print *, ' or '
     print *, "   fcompare file1 file2"
     print *, " "
     stop
  endif

  !---------------------------------------------------------------------------
  ! build the plotfiles and do initial comparisons
  
  unit_a = unit_new()
  call build(pf_a, plotfile_a, unit_a)

  unit_b = unit_new()
  call build(pf_b, plotfile_b, unit_b)

  dm = pf_a%dim
  
  ! check if they are the same dimensionality
  if (pf_a%dim /= pf_b%dim) then
     call bl_error("ERROR: plotfiles have different numbers of spatial dimensions")
  endif


  ! check if they have the same number of levels
  if (pf_a%flevel /= pf_b%flevel) then
     call bl_error("ERROR: number of levels do not match")
  endif


  ! check if the finest domains are the same size
  bx_a = plotfile_get_pd_box(pf_a, pf_a%flevel)
  flo_a = 1
  fhi_a = 1
  flo_a(1:dm) = lwb(bx_a)
  fhi_a(1:dm) = upb(bx_a)

  bx_b = plotfile_get_pd_box(pf_b, pf_b%flevel)
  flo_b(1:dm) = lwb(bx_b)
  fhi_b(1:dm) = upb(bx_b)

  if ( (flo_a(1) /= flo_b(1) .OR. fhi_a(1) /= fhi_b(1)) .OR. &
      ((flo_a(2) /= flo_b(2) .OR. fhi_a(2) /= fhi_b(2)) .AND. pf_a%dim >= 2) .OR. &
      ((flo_a(3) /= flo_b(3) .OR. fhi_a(3) /= fhi_b(3)) .AND. pf_a%dim == 3) ) then
     call bl_error("ERROR: grids do not match")
  endif

  
  ! check if they have the same number of variables
  if (pf_a%nvars /= pf_b%nvars) then
     print *, " "
     print *, "WARNING: number of variables do not match"
  endif

  allocate(aerror(pf_a%nvars))
  allocate(rerror(pf_a%nvars))

  allocate(has_nan_a(pf_a%nvars))
  allocate(has_nan_b(pf_a%nvars))

  any_nans = .false.

  ! in case the variables are not in the same order, figure out the
  ! mapping between pf_a and pf_b variables
  allocate(ivar_b(pf_a%nvars))

  do n_a = 1, pf_a%nvars

     ivar_b(n_a) = -1
     do n_b = 1, pf_b%nvars

        if (pf_a%names(n_a) == pf_b%names(n_b)) then
           ivar_b(n_a) = n_b
           exit
        endif

     enddo

     if (ivar_b(n_a) == -1) then
        print *, "WARNING: variable ", trim(pf_a%names(n_a)), &
                 " not found in plotfile 2"
     endif

  enddo


  ! also print out, as a diagnostic, those variables in plotfile 1 that
  ! are not in plotfile 2
  do n_b = 1, pf_b%nvars
     itest = -1
     do n_a = 1, pf_a%nvars

        if (pf_a%names(n_a) == pf_b%names(n_b)) then
           itest = n_a
           exit
        endif

     enddo

     if (itest == -1) then
        print *, "WARNING: variable ", trim(pf_b%names(n_b)), &
                 " not found in plotfile 1"
     endif

  enddo

  !---------------------------------------------------------------------------
  ! go level-by-level and patch-by-patch and compare the data
  
998 format(1x,a24,2x,a20,   2x,a20)
999 format(1x,70("-"))

  write (*,*) " "
  write (*,998) "variable name", "absolute error", "relative error"
  write (*,999)

  do i = 1, pf_a%flevel

     aerror(:) = ZERO
     rerror(:) = ZERO

     has_nan_a(:) = .false.
     has_nan_b(:) = .false.

     ! make sure the dx's agree
     dx_a = 0.0_dp_t
     dx_b = 0.0_dp_t
     dx_a(1:dm) = plotfile_get_dx(pf_a, i)
     dx_b(1:dm) = plotfile_get_dx(pf_b, i)  

     if ((dx_a(1) /= dx_b(1)) .OR. &
         (pf_a%dim >= 2 .AND. dx_a(2) /= dx_b(2)) .OR. &
         (pf_a%dim == 3 .AND. dx_a(3) /= dx_b(3))) then
        call bl_error("ERROR: grid dx does not match")
     endif

     ! make sure the number of boxes agree
     nboxes_a = nboxes(pf_a, i)
     nboxes_b = nboxes(pf_b, i)
     
     if (nboxes_a /= nboxes_b) then
        call bl_error("ERROR: number of boxes do not match")
     endif

     do j = 1, nboxes_a

        ! make sure that the grids match
        bx_a = get_box(pf_a, i, j)
        lo_a = 1
        hi_a = 1
        lo_a(1:dm) = lwb(bx_a)
        hi_a(1:dm) = upb(bx_a)

        bx_b = get_box(pf_b, i, j)
        lo_b = 1
        hi_b = 1
        lo_b(1:dm) = lwb(bx_b)
        hi_b(1:dm) = upb(bx_b)

        if ( (lo_a(1) /= lo_b(1) .OR. hi_a(1) /= hi_b(1)) .OR. &
             (pf_a%dim >= 2 .AND. (lo_a(2) /= lo_b(2) .OR. hi_a(2) /= hi_b(2))) .OR. &
             (pf_a%dim == 3 .AND. (lo_a(3) /= lo_b(3) .OR. hi_a(3) /= hi_b(3))) ) then
           call bl_error("ERROR: grids do not match")
        endif

        ! loop over the variables.  Take plotfile_a to be the one defining
        ! the list of variables, and bind them one-by-one.  Don't assume that
        ! the variables are in the same order in plotfile_b.

        do n_a = 1, pf_a%nvars

           n_b = ivar_b(n_a)
           if (n_b == -1) cycle

           call fab_bind_comp_vec(pf_a, i, j, (/ n_a /) )
           call fab_bind_comp_vec(pf_b, i, j, (/ n_b /) )

           p_a => dataptr(pf_a, i, j)
           p_b => dataptr(pf_b, i, j)

           ! check for NaNs -- comparisons don't work when they are present
           call fab_contains_nan(p_a, &
                                 (hi_a(3)-lo_a(3)+1)* &
                                 (hi_a(2)-lo_a(2)+1)* &
                                 (hi_a(1)-lo_a(1)+1),  ir_a)

           if (ir_a == 1) has_nan_a(n_a) = .true.


           call fab_contains_nan(p_b, &
                                 (hi_b(3)-lo_b(3)+1)* &
                                 (hi_b(2)-lo_b(2)+1)* &
                                 (hi_b(1)-lo_b(1)+1),  ir_b)

           if (ir_b == 1) has_nan_b(n_a) = .true.

           if (has_nan_a(n_a) .or. has_nan_b(n_a)) cycle


           do kk = lo_a(3), hi_a(3)
              do jj = lo_a(2), hi_a(2)
                 do ii = lo_a(1), hi_a(1)

                    pa = abs(p_a(ii,jj,kk,1))
                    pb = abs(p_b(ii,jj,kk,1))
                    pd = abs(p_a(ii,jj,kk,1) - p_b(ii,jj,kk,1))

                    if (norm == 0) then
                       aerror(n_a) = max(aerror(n_a),pd)
                       if (pa .ne. 0.d0) then 
                          rerror(n_a) = max(rerror(n_a), pd/pa)
                       else if (pb .ne. 0.d0) then 
                          rerror(n_a) = max(rerror(n_a), pd/pb)
                       else 
                          ! The relative error is zero so do nothing
                       end if
                    else
                       aerror(n_a) = aerror(n_a) + pd**norm
                       if (pa .ne. 0.d0) then 
                           rerror(n_a) = rerror(n_a) + (pd/pa)**norm
                       else if (pb .ne. 0.d0) then 
                           rerror(n_a) = rerror(n_a) + (pd/pb)**norm
                       else 
                          ! The relative error is zero so do nothing
                       end if
                    endif

                 enddo
              enddo
           enddo

           call fab_unbind(pf_a, i, j)
           call fab_unbind(pf_b, i, j)

        enddo  ! variable loop

     enddo  ! boxes loop

     ! Normalize
     if (norm > 0) then

        do n_a = 1, pf_a%nvars
           aerror(n_a) = aerror(n_a)*product(dx_a(1:pf_a%dim))
           aerror(n_a) = aerror(n_a)**(ONE/real(norm,dp_t))

           rerror(n_a) = rerror(n_a)*product(dx_a(1:pf_a%dim))
           rerror(n_a) = rerror(n_a)**(ONE/real(norm,dp_t))
        enddo

     endif


     !------------------------------------------------------------------------
     ! print out the comparison report for this level
     
1000 format(1x,"level = ", i2)
1001 format(1x,a24,2x,g20.10,2x,g20.10)
1002 format(1x,a24,2x,a42)

     write (*,1000) i

     do n_a = 1, pf_a%nvars
        if (ivar_b(n_a) == -1) then
           write (*,1002) pf_a%names(n_a), "< variable not present in both files > "
        else if (has_nan_a(n_a) .or. has_nan_b(n_a)) then
           write (*,1002) pf_a%names(n_a), "< NaN present > "
        else
           write (*,1001) pf_a%names(n_a), aerror(n_a), rerror(n_a)
        endif
     enddo

     if (i == 1) then
        global_error = maxval(aerror(:))
     else
        global_error = max(global_error, maxval(aerror(:)))
     endif

     any_nans = any(has_nan_a .or. has_nan_b) .or. any_nans

  enddo  ! level loop

  if (global_error == ZERO .and. .not. any_nans) then
     print *, "PLOTFILES AGREE"
  endif

end program fcompare

File: Tools/Postprocessing/C_Src/fcompare.cpp
//
// C++ version of fcompare.f90
//
// Take 2 plotfiles as input and compare each level zone by zone for differences.
// The grids at each level may not be identical, but must have the same
// problem domain.
//
// This does not rely on FBaselib.
//
//  Author: Michele Rosso
//  Date  : May, 2018
//
#include "AMReX_DataServices.H"

using namespace amrex;
using namespace std;

//
// Prototypes
//
void GetInputArgs ( const int argc, char** argv,
		    bool& ghost, int& norm,
		    string& diffvar, string& zonevar,
		    string& file1, string& file2 );

void PrintHelp ( const char* progName );


//
//
//
int main ( int argc, char* argv[] )
{

    // Use "false" as third arguments to prevent AMReX
    // to check for an input file as first command line argument
    amrex::Initialize ( argc, argv, false );
    {

    // This runs in serial for now
    if ( ParallelDescriptor::NProcs() > 1 )
    	Abort("Only run this with one processor");

    // Input arguments
    bool    do_ghost=false;
    int     norm=0;
    string  diffVar;
    string  zoneVar;
    string  file1;
    string  file2;


    GetInputArgs ( argc, argv, do_ghost, norm, diffVar, zoneVar,
		   file1, file2 );

    // Start dataservices (no clue why we need to do this )
    DataServices::SetBatchMode();

    // Define the type of file
    Amrvis::FileType fileType(Amrvis::NEWPLT);
    DataServices dataServices1 (file1, fileType);
    DataServices dataServices2 (file2, fileType);

    // NOTE:
    // fcompare.f90 checks that both files have the same dimensionality
    // and the dimensionality s the same as AMREX_SPACEDIM used to build fcompare itself.
    // fcompare.cpp does not need to do this since dataservices take care
    // of checking that when loading the files.
    if ( !dataServices1.AmrDataOk() || !dataServices2.AmrDataOk() )
	DataServices::Dispatch(DataServices::ExitRequest, NULL);

    // get data from plot files
    AmrData& data1 = dataServices1.AmrDataRef();
    AmrData& data2 = dataServices2.AmrDataRef();

    // Check if both plotfiles have the same numbers of levels
    if ( data1.FinestLevel () != data2.FinestLevel () )
	Abort ("number of levels do not match");

    int finestLevel = data1.FinestLevel ();

    // Check that the problem domain (bounding box) are the same size
    // check that boxarray and distribution maps are the same
    // for both plotfiles
    const Vector<Box>& bb1 = data1.ProbDomain ();
    const Vector<Box>& bb2 = data2.ProbDomain ();

    for (int nl = 0; nl <= finestLevel; ++nl )
    {
	if ( bb1[nl] != bb2[nl] )
    	    Abort ("grids do no match");

	const BoxArray& ba1 = data1.boxArray(nl);
	const BoxArray& ba2 = data2.boxArray(nl);
	const DistributionMapping& dm1 = data1.DistributionMap (nl);
	const DistributionMapping& dm2 = data2.DistributionMap (nl);

	AMREX_ALWAYS_ASSERT ( ba1 == ba2 );
	AMREX_ALWAYS_ASSERT ( dm1 == dm2 );
    }

    // Check if plotfiles have same numbers of variables
    if ( data1.NComp () != data2.NComp () )
	Warning ("WARNING: number of variables do not match");

    // Check grid spacing is the same for both plotfiles
    const Vector<Vector<Real> >& dxLevel1 = data1.DxLevel();
    const Vector<Vector<Real> >& dxLevel2 = data2.DxLevel();

    for (int nl = 0; nl <= finestLevel; ++nl )
	for (int nd = 0; nd < AMREX_SPACEDIM; ++nd )
	    if ( dxLevel1[nl][nd] != dxLevel2[nl][nd] )
		Abort ("grid dx does not match");

    // Check for the same number of ghost cells
    int ng = 0;

    if ( data1.NGrow() != data2.NGrow() )
    {
	Warning ("grids have different numbers of ghost cells");
    }
    else
    {
	if (do_ghost)
	    ng = data1.NGrow();
    }

    // Check if variables in file 1 are also in file 2
    // and print a warning if this is not the case
    const Vector<string>& varNames1 = data1.PlotVarNames ();
    const Vector<string>& varNames2 = data2.PlotVarNames ();

    bool         allVarsFound = true;
    Vector<int>  varMap(data1.NComp(),-1);

    for (int nv2 = 0; nv2 < data2.NComp (); ++nv2 )
    {
	bool varFound = false;

	for (int nv1 = 0; nv1 < data1.NComp (); ++nv1 )
	{
	    if ( varNames2[nv2] == varNames1[nv1] )
	    {
		varMap[nv1] = nv2;
		varFound = true;
		break;
	    }
	}


	if ( !varFound )
	{
	    string warnmsg = "WARNING: variable " + varNames2[nv2]
		+ " not found in plotfile " + file1;

	    Warning (warnmsg);

	    allVarsFound = false;
	}

    }

    //
    // Go level-by-level and patch-by-patch and compare the data
    //
    cout << endl;

    cout << setw(25) << " variable name  "
    	 << setw(25) << " absolute error "
    	 << setw(25) << " relative error "
    	 << endl;

    cout << setw(25) << " "
    	 << setw(22) << "(||A - B||)"
    	 << setw(29) << "(||A - B||/||A||)"
    	 << endl;

    cout << string(77, '-') << std::endl;

    vector<Real>  aerror(data1.NComp());
    vector<Real>  rerror(data1.NComp());
    vector<Real>  rerror_den(data1.NComp());
    vector<bool>  has_nan(data1.NComp());
    Real          globalError = 0.0;
    bool          anyNaNs = false;

    // Loop over levels
    for (int nl = 0; nl <= finestLevel; ++nl )
    {

	// Zero arrays
	fill ( aerror.begin(),     aerror.end(), 0.0 );
	fill ( rerror.begin(),     rerror.end(), 0.0 );
	fill ( rerror_den.begin(), rerror_den.end(), 0.0 );
	fill ( has_nan.begin(),   has_nan.end(), false );

	// We already checked that boxarray is the same for
	// both plotfiles
	const BoxArray& ba = data1.boxArray(nl);

	// Loop over boxes on same level
    	for (int nb = 0; nb < ba.size (); ++nb )
    	{
	    // Create a Multifab based on a boxarray with only
	    // one box.
	    // This way we only have in memory a box at a time, but we can
	    // still use the capabilities of MultiFab  class
	    BoxArray             sba(ba[nb]); // Single Box Array
	    DistributionMapping  dm(sba);
	    MultiFab             mf1( sba, dm, 1, ng );
	    MultiFab             mf2( sba, dm, 1, ng );
	    MultiFab             mfdiff( sba, dm, 1, ng );

            //
	    // In doing the calculation below, assume that the
	    // data are always cell-centered
	    //
	    for (int nv = 0; nv < data1.NComp(); ++nv)
	    {
		if ( varMap[nv] == -1 )
		    continue;

		data1.FillVar ( mf1, nl, varNames1[nv] );
		data2.FillVar ( mf2, nl, varNames2[varMap[nv]] );

		// check for NaNs -- comparisons don't work when they are present
		// note: regardless of do_ghost, this will check the ghostcells
		// too if they are present.
		if ( mf1.contains_nan() || mf2.contains_nan() )
		{
		    has_nan[nv] = true;
		    anyNaNs = true;
		    continue;
		}

		MultiFab::LinComb ( mfdiff, 1.0, mf1, 0, -1.0, mf2, 0, 0, 1, ng );

		if ( norm == 0 )
		{
		    aerror[nv]       = amrex::max ( mfdiff.norm0(0,ng), aerror[nv] );
		    rerror_den[nv]   = amrex::max ( mf1.norm0(0,ng),    rerror_den[nv] );
		}
		else if ( norm == 1 )
		{
		    aerror[nv]       = aerror[nv] + mfdiff.norm1(0,ng);
		    rerror_den[nv]   = rerror_den[nv] + mf1.norm1();
		}
		else if ( norm == 2 )
		{
		    aerror[nv]       = aerror[nv] + MultiFab::Dot ( mfdiff, 0, mfdiff, 0, 1, ng );
		    rerror_den[nv]   = rerror_den[nv] + MultiFab::Dot ( mf1, 0, mf1, 0, 1, ng );
		}

		rerror[nv] = aerror[nv];

    	    }

	    data1.FlushGrids();
	    data2.FlushGrids();
    	}

	// Compute normalized norms for this level and print out value
	// What  if nghost are included?
	// In this case the normalization by grid volume is not correct
	if ( norm > 0 )
	{
	    // Metric coefficient
	    Real dvol = 1.0;

	    for (int nd = 0; nd < AMREX_SPACEDIM; ++nd )
		dvol = dvol * dxLevel1[nl][nd];


	    for (int nv = 0; nv < data1.NComp(); ++nv )
	    {
		aerror[nv]  = pow ( aerror[nv]*dvol, 1.0/norm );
		rerror[nv]  = aerror[nv] / pow ( rerror_den[nv]*dvol, 1.0/norm );
	    }
	}
	else
	{
	    for (int nv = 0; nv < data1.NComp(); ++nv )
		rerror[nv]  = aerror[nv] / rerror_den[nv];
	}

	//
	// print out the comparison report for this level
	//
	Print() << " level = " << setw(2) << nl << endl;

	for (int nv = 0; nv < data1.NComp(); ++nv )
	{
	   if ( varMap[nv] == -1 )
	   {
	       Print() << " " << setw(22) << varNames1[nv] << "  "
		       << setw(50) << "< variable not present in both files > "
		       << endl;
	   }
	   else if ( has_nan[nv] == true )
	   {
	       Print() << " " << setw(24) << varNames1[nv] << "  "
	   	       << setw(50) << "< NaN present > "
	   	       << endl;
	   }
	   else
	   {
	       Real aerr = 0.0;
	       Real rerr = 0.0;

	       if ( aerror[nv] > 0.0 )
		   aerr = amrex::min ( amrex::max ( aerror[nv], 1e-99 ), 1e+98 );

	       if ( rerror[nv] > 0.0 )
		   rerr = amrex::min ( amrex::max ( rerror[nv], 1e-99 ), 1e+98 );

	       Print() << scientific;
	       Print() << " " << setw(22) << varNames1[nv] << "  "
		       << setw(24) << setprecision(10)  << uppercase << aerr
		       << "  " << setw(24) << setprecision(10)  << rerr
		       << endl;
	   }


	}

	// compute global error
	for ( unsigned int nv = 0; nv < aerror.size(); ++nv )
	    globalError = amrex::max (globalError, aerror[nv] );

    }

    if ( globalError == 0.0 && !anyNaNs)
    {
	if ( !allVarsFound )
	    Warning ("WARNING: not all variables present in both files");

	Print() << "PLOTFILES AGREE" << endl;

	exit (EXIT_SUCCESS);
    }
    else
    {
	exit (EXIT_FAILURE);
    }

    }
    Finalize ();
}



//
// Parse command line arguments
//
void GetInputArgs ( const int argc, char** argv,
		    bool& ghost, int& norm,
		    string& diffvar, string& zonevar,
		    string& file1, string& file2 )
{

    int i = 1; // skip program name

    while ( i < argc - 2 ) // Skip last two: those are the files names
    {

	if ( !strcmp(argv[i],"-g") || !strcmp(argv[i],"--ghost") )
	{
	    ghost = true;
	}
	else if ( !strcmp(argv[i],"-n") || !strcmp(argv[i],"--norm") )
	{
	    norm = atoi(argv[++i]);

	    if ( ( norm < 0 ) || ( norm > 2) )
	    {
		cout << "\nNorm type is " << norm << endl;
		cout << "Valid values are 0, 1, 2. " << endl;
		PrintHelp ( argv[0] );
		exit (EXIT_FAILURE);
	    }

	}
	else if ( !strcmp(argv[i],"-d") || !strcmp(argv[i],"--diffvar") )
	{
	    Warning ("Option -d/--diffvar is not supported");
	    //diffvar = argv[++i];
	}
	else if ( !strcmp(argv[i],"-z") || !strcmp(argv[i],"--zone_info") )
	{
	    Warning ("Option -z/--zone_info is not supported");
	    //zonevar = argv[++i];
	}
	else
	{
	    std::cout << "\n\nOption " << argv[i] << " not recognized" << std::endl;
	    PrintHelp ( argv[0] );
	    exit ( EXIT_FAILURE );
	}

	// Go to the next parameter name
	++i;
    }

    // get name of input files
    if (argc > 2)
    {
	file1 = argv[argc-2];
	file2 = argv[argc-1];
    }

    if ( file1.empty() || file2.empty() )
    {
	PrintHelp (argv[0]);
	Abort("Missing one or both input files");
    }

    // Print summary
    cout << "\nFile 1 is " + file1
	 << "\nFile 2 is " + file2;

    if (ghost)
	cout << "\nGhost cells will be compared (if stored)";

    if (!diffvar.empty())
	cout << "\nPlotfile with differences for variable " + diffvar
	     << " will be output";

    if (!zonevar.empty())
	cout << "\nInfo on max error zone for var " + zonevar
	     << " will be given";

    cout << endl << endl;
}



//
// Print usage info
//
void PrintHelp (const char* progName)
{
    cout  << "\nCompare two plotfiles, zone by zone, to machine precision"
	  << "\nand report the maximum absolute and relative errors for each"
	  << "\nvariable."
	  << "\n "
	  << "\nUsage:"
	  << "\n " << progName << " [-g|--ghost] [-n|--norm num] file1 file2"
	// << "\n " << progName << " [-g|--ghost] [-n|--norm num] [--diffvar var] [-z|--zone_info var] file1 file2"
	  << "\n "
	  << "\nOptional arguments:"
	  << "\n   -g|--ghost         : compare the ghost cells too (if stored)"
	  << "\n   -n|--norm num      : what norm to use (default is 0 for inf norm)"
	  // << "\n   -d|--diffvar var   : output a plotfile showing the differences for "
	  // << "\n                        variable var"
	  // << "\n   -z|--zone_info var : output the information for a zone corresponding"
	  // << "\n                        to the maximum error for the given variable"
	  << "\n\n" << endl;

}
